compiler/rustc_const_eval/src/interpret/validity.rs

```diff
@@ -724,20 +724,21 @@ fn mutability<'mir, 'tcx: 'mir>(
                 // so just use the declared mutability.
                 mutability
             } else {
+                let ty = ecx
+                    .tcx
+                    .type_of(did)
+                    .no_bound_vars()
+                    .expect("statics should not have generic parameters");
                 let mutability = match mutability {
-                    Mutability::Not
-                        if !ecx
-                            .tcx
-                            .type_of(did)
-                            .no_bound_vars()
-                            .expect("statics should not have generic parameters")
-                            .is_freeze(*ecx.tcx, ty::ParamEnv::reveal_all()) =>
-                    {
+                    Mutability::Not if !ty.is_freeze(*ecx.tcx, ty::ParamEnv::reveal_all()) => {
                         Mutability::Mut
                     }
                     _ => mutability,
                 };
-                if let Some((_, alloc)) = ecx.memory.alloc_map.get(alloc_id) {
+                if let Some((_, alloc)) = ecx.memory.alloc_map.get(alloc_id)
+                    // For type errors, we do not know whether they are supposed to be mutable or not.
+                    && !ty.references_error()
```

Comment:
It is a bad sign that this check is needed. We should never even evaluate a static whose type references an error. So something still seems wrong here.